docs/configuration.md
# Configuration

Before start configuring, make sure that the config setting 
``ckan.saml_custom_base_path`` is set if your config file is not stored at
``/etc/ckan/default/saml`` directory.

Copy the ``saml`` folder from ``ckanext-saml`` into the directory where your 
CKAN config file is placed:

```
cp -R saml_example/saml YOUR_CONFIG_DIRECTORY/saml
```

Open the ``settings.json`` file that is in your copied ``saml`` folder in order 
to modify it.

You can also set them in the CKAN configuration file (e.g. ckan.ini or 
production.ini) using with prefix ``ckanext.saml.settings.substitution`` 
or using the admin configuration page.


## Configure main settings file

The main sections that should be updated within the file are``sp`` and ``idp``.

### SP modifications (CKAN):

1. Modify ``entityId`` with the domain name of your portal.

2. Modify ``url`` in ``assertionConsumerService`` with the domain name of your
   portal plus adding ``/saml/`` at the end. This is the URL where IdP will
   send back the reponse with User Data. Make sure the the slash is present in
   the end of the URL.

### IdP modifications

1. Modify ``entityId`` by filling it with the ``entityID`` that should be 
present in the ``idp_metadata.xml`` file (name of the file can be different) 
that is been sent to you by the IdP.

2. Modify ``url`` in ``singleSignOnService``. You can find this ``url`` in 
``idp_metadata.xml`` at ``SingleSignOnService`` section, it should have 
``Location`` attribute where the url is specified.

3. Modify ``x509cert`` by filling it with the ``X509Certificate`` that should be 
present in ``idp_metadata.xml``. Make sure this set as a **single line string**.

**Note**:  ``singleLogoutService`` is not implemented.


### Other modifications

``advanced_settings.json`` is used for additional configurations such as
security.  It also needed to modify the ``contactPerson`` and ``organization``
sections in it if your are going to provide your SP data to your IdP.

After updating all mentioned values in ``settings.json``, at
``DOMAIN_NAME/saml/metadata URL`` you can find the ``sp.xml`` generated, which
you can provide to the IdP for configuration on their side.

The main infomation that is needed for the IdP is the
``AssertionConsumerService``(ACS) which should be set on their APP for
SAML. ``AssertionConsumerService`` should match to what you have in your
settings.json and IdP APP, otherwise errors might appear.


## Configuration options

There is a number of configuration options available for the extension. You can 
set them in the CKAN configuration file (e.g. ckan.ini or production.ini).


### Config SSO path

**`ckanext.saml.sso_path`** [__optional__]

Set ``path`` to single sign-on.

**Type**: `str`

**Default**: `/sso/post`

-----

### Config SLO path

**`ckanext.saml.slo_path`** [__optional__]

Set ``path`` to single logout.

**Type**: `str`

**Default**: `/slo/post`

-----

### Config static host

**`ckanext.saml.static_host`** [__optional__]

Set the name of static host for SSO.

**Type**: `str`

**Default**: `None`

-----

### IdP metadata url 

**`ckanext.saml.metadata.url`** [__optional__]

Set the URL to IdP remote metadata.

**Type**: `str`

**Default**: `None`

-----

### Login button text

**`ckanext.saml.login_button_text`** [__optional__]

Provides an ability to customize login button text.

**Legacy**: `ckan.saml_login_button_text`

**Type**: `str`

**Default**: `SAML Login`

-----

### Mapper filename

**`ckan.saml_custom_attr_map`** [__optional__]

Used to modify mapper filename.

**Type**: `str`

**Default**: `mapper.py`

-----

### NameID as an email

**`ckan.saml_use_nameid_as_email`** [__optional__]

Set to ``true`` if you want to use NameID as an email for the User in order not 
to claim it additionally from the IdP.

**Type**: `bool`

**Default**: `false`

-----

### Name from Response

**`ckan.saml.name_from_response`** [__optional__]

Set to ``true`` if you want the User name be taken from the Response instead of generating random one.

**Type**: `bool`

**Default**: `false`

-----

### Fields trigger User update

**`ckan.saml.user_fields_trigger_update`** [__optional__]

By default, when User is logging in, the User update action is called only in case if ``fullname`` of the User from the response is updated. To extended this list, you can provide fields separated by whitespace, for example  ``fullname name givename``.

**Type**: `string`

**Default**: `fullname`

-----

### Path to error TPL

**`ckanext.saml.error_template`** [__optional__]

Set ``path`` to custom template for errors rendering.

**Type**: `str`

**Default**: `None`

-----

### Path to SAML config files

**`ckanext.saml.metadata.base_path`** [__optional__]

Provides custom path where saml settings files/folders will be searched.

**Legacy**: `ckan.saml_custom_base_path`

**Type**: `str`

**Default**: `/etc/ckan/default/saml`

-----

### Reactivate deleted account

**`ckanext.saml.reactivate_deleted_account`** [__optional__]

Change the state of a `deleted` account status to `active`.

**Type**: `bool`

**Default**: `false`

-----

### RelayState path

**`ckan.saml_relaystate`** [__optional__]

Set a custom RelayState ``path``.

**Type**: `str`

**Default**: `None`

-----

### Session TTL

**`ckanext.saml.session.ttl`** [__optional__]

Set the time a user can remain idle before the session is terminated and the 
user must log in again.

**Type**: `str`

**Default**: `30 * 24 * 3600`

-----

### Use HTTPS

**`ckan.saml_use_https`** [__optional__]

Used to send data while **https**, set ``on`` to enable it.

**Options**: `on`, `off`

**Type**: `bool`

**Default**: `off`

-----

### Using ``ckan.root_path``

**`ckan.saml_use_root_path`** [__optional__]

This needs to be set to ``true`` if you run your portal using the 
``ckan.root_path``.

**Type**: `bool`

**Default**: `false`

-----

### Use dynamic

**`ckanext.saml.settings.dynamic`** [__optional__]

Use dynamic Single Sign-On (SSO) URLs rather than pre-defined static SSO URLs.

**Type**: `bool`

**Default**: `false`

-----

### Use remote IdP

**`ckanext.saml.metadata.remote_idp`** [__optional__]

Use remote identity provider.

**Type**: `bool`

**Default**: `false`

-----

### Use forwarded host

**`ckanext.saml.use_forwarded_host`** [__optional__]

Use HTTP_X_FORWARDED_HOST when there is a proxy (or multiple proxies) between 
the browser and your server.

**Type**: `bool`

**Default**: `false`

-----

### Use unconditional login

**`ckanext.saml.unconditional_login`** [__optional__]

Use unconditional login for single sign-on.

**Type**: `bool`

**Default**: `false`

-----


## SP metadata

As mentioned above, you can find SP metadata at ``DOMAIN_NAME/saml/metadata
URL`` after configuring ``advanced_settings.json``.  This **URL** is accessible
only to ``sysadmins`` and presented in **XML** format.  Additional tab on
``/ckan-admin/`` is added, that leads to this page.


## Data encryption

In order to encrypt the coming data from the IdP use ``advanced_settings.json``
file. In ``security`` section, you can enable encryption for NameId and all
other data that will be returned to the SP.

If you enable one of
``authnRequestsSigned``,``logoutRequestSigned``,``logoutResponseSigned``,
``wantAssertionsEncrypted``, ``wantNameIdEncrypted`` (you can find description 
of each option [here](https://github.com/onelogin/python3-saml#how-it-works)), 
you will have to create [x509 certificate](https://en.wikipedia.org/wiki/X.509) 
in you SP. Cerificate should be created in ``certs`` folder, files should be 
named as ``sp.crt`` and ``sp.key`` (private key). After creating it, your 
``sp.xml`` will show you public key ``ds:X509Certificate`` that should be 
delivered to your IdP in order to configure encryption.


## Extras

ckanext-saml has interface ``ICKANSAML`` which has two hooks that can be used
for User data modificaiton and Organization memberships logic while login.

- ``after_mapping`` - Used after Users data is being mapped, but before the
User is being created.

- ``roles_and_organizations`` - Used for adding custom logic for Organization
membership that is going to be applied to the User. There is no default logic
for this, so should be added in your custom extension using this hook.


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](http://keepachangelog.com/en/1.0.0/)
and this project adheres to [Semantic Versioning](http://semver.org/spec/v2.0.0.html).

<!-- insertion marker -->
## [v0.3.6](https://github.com/DataShades/ckanext-saml/releases/tag/v0.3.6) - 2025-02-04

<small>[Compare with v0.3.5](https://github.com/DataShades/ckanext-saml/compare/v0.3.5...v0.3.6)</small>

### Features

- add ckan.saml.name_from_response and ckan.saml.user_fields_trigger_update options ([3f2e0e2](https://github.com/DataShades/ckanext-saml/commit/3f2e0e2960db4e30b14c130b6cb3fcff5ec887dc) by Sergey Motornyuk).

## [v0.3.5](https://github.com/DataShades/ckanext-saml/releases/tag/v0.3.5) - 2024-12-16

<small>[Compare with v0.3.4](https://github.com/DataShades/ckanext-saml/compare/v0.3.4...v0.3.5)</small>

### Features

- add saml_before_user_create method to ICKANSAML interface ([574cea7](https://github.com/DataShades/ckanext-saml/commit/574cea7e46c947bb7df57c00934335d0861b2d1e) by Sergey Motornyuk).

## [v0.3.4](https://github.com/DataShades/ckanext-saml/releases/tag/v0.3.4) - 2024-09-03

<small>[Compare with v0.3.3](https://github.com/DataShades/ckanext-saml/compare/v0.3.3...v0.3.4)</small>

### Bug Fixes

- do not duplicate dynamic views ([d53ce75](https://github.com/DataShades/ckanext-saml/commit/d53ce75f086e159ddaab5c9f951b1977951551c1) by Sergey Motornyuk).

## [v0.3.3](https://github.com/DataShades/ckanext-saml/releases/tag/v0.3.3) - 2024-04-18

<small>[Compare with v0.3.2](https://github.com/DataShades/ckanext-saml/compare/v0.3.2...v0.3.3)</small>

### Features

- remove overrides for user edit form ([3205480](https://github.com/DataShades/ckanext-saml/commit/3205480bc001a8045825361834537954b8495791) by Sergey Motornyuk).

## [v0.3.2](https://github.com/DataShades/ckanext-saml/releases/tag/v0.3.2) - 2023-09-07

<small>[Compare with v0.3.1](https://github.com/DataShades/ckanext-saml/compare/v0.3.1...v0.3.2)</small>

### Features

- read remote metadata from file ([b912844](https://github.com/DataShades/ckanext-saml/commit/b9128446c2e713adb7053a2c056628efebb68940) by Sergey Motornyuk).
- saml_idp_refresh accepts `url` parameter ([d66989b](https://github.com/DataShades/ckanext-saml/commit/d66989b9d9a71c07b7ea2260be0ceea86b65fb2d) by Sergey Motornyuk).

### Bug Fixes

- automatically call idp_refresh when it is not available in cache ([3cc8291](https://github.com/DataShades/ckanext-saml/commit/3cc82912414403432f9e7a7a97f9928f421474ff) by Sergey Motornyuk).

## [v0.3.1](https://github.com/DataShades/ckanext-saml/releases/tag/v0.3.1) - 2023-04-24

<small>[Compare with v0.3.0](https://github.com/DataShades/ckanext-saml/compare/v0.3.0...v0.3.1)</small>

### Bug Fixes

- add csrf_token to user form ([05b0048](https://github.com/DataShades/ckanext-saml/commit/05b0048b91e96560ef419fc902402897da65979a) by Sergey Motornyuk).

## [v0.3.0](https://github.com/DataShades/ckanext-saml/releases/tag/v0.3.0) - 2023-04-03

<small>[Compare with v0.2.1](https://github.com/DataShades/ckanext-saml/compare/v0.2.1...v0.3.0)</small>

### Features

- CKAN v2.10 support ([6f0fb3d](https://github.com/DataShades/ckanext-saml/commit/6f0fb3d1a0a07eb842e24d9cae78d8561d8f5084) by Sergey Motornyuk).

## [v0.2.1](https://github.com/DataShades/ckanext-saml/releases/tag/v0.2.1) - 2023-02-02

<small>[Compare with v0.2.0](https://github.com/DataShades/ckanext-saml/compare/v0.2.0...v0.2.1)</small>

### Bug Fixes

- treat `ckan.saml_use_nameid_as_email` as bool ([282cbf1](https://github.com/DataShades/ckanext-saml/commit/282cbf19430bd1873ec15fddfa7f9ffe1d8197e5) by Sergey Motornyuk).

## [v0.2.0](https://github.com/DataShades/ckanext-saml/releases/tag/v0.2.0) - 2022-11-10

<small>[Compare with v0.1.8](https://github.com/DataShades/ckanext-saml/compare/v0.1.8...v0.2.0)</small>

### Bug Fixes

- additional fixes for handling removed users ([421c948](https://github.com/DataShades/ckanext-saml/commit/421c948e4ae13132dd9ea5dcc95a960c271d043c) by Sergey Motornyuk).
- use came_from ([1c768cc](https://github.com/DataShades/ckanext-saml/commit/1c768cc3b6ad95a703046170a69f835a25b8e182) by Sergey Motornyuk).
- reactivate deleted accounts ([7ae982d](https://github.com/DataShades/ckanext-saml/commit/7ae982d91923696fbe96179ea00f850736235b14) by Sergey Motornyuk).

## [v0.1.8](https://github.com/DataShades/ckanext-saml/releases/tag/v0.1.8) - 2022-11-09

<small>[Compare with v0.1.7](https://github.com/DataShades/ckanext-saml/compare/v0.1.7...v0.1.8)</small>

## [v0.1.7](https://github.com/DataShades/ckanext-saml/releases/tag/v0.1.7) - 2022-11-09

<small>[Compare with v0.1.6](https://github.com/DataShades/ckanext-saml/compare/v0.1.6...v0.1.7)</small>

### Bug Fixes

- configurable redirect after login ([bfea66b](https://github.com/DataShades/ckanext-saml/commit/bfea66b17ed8398108ba1f82279f6a280063d18e) by Sergey Motornyuk).

## [v0.1.6](https://github.com/DataShades/ckanext-saml/releases/tag/v0.1.6) - 2022-10-25

<small>[Compare with v0.1.5](https://github.com/DataShades/ckanext-saml/compare/v0.1.5...v0.1.6)</small>

### Features

- Allow login without sso URL parameter ([3d78e2a](https://github.com/DataShades/ckanext-saml/commit/3d78e2ad5391575c5d76d6c936b07ba639e2c3d9) by Sergey Motornyuk).

## [v0.1.5](https://github.com/DataShades/ckanext-saml/releases/tag/v0.1.5) - 2022-10-24

<small>[Compare with v0.1.4](https://github.com/DataShades/ckanext-saml/compare/v0.1.4...v0.1.5)</small>

### Bug Fixes

- standardize config option names ([108db0f](https://github.com/DataShades/ckanext-saml/commit/108db0f31e336d3b38986e5b2e26ca01553d5dc7) by Sergey Motornyuk).

## [v0.1.4](https://github.com/DataShades/ckanext-saml/releases/tag/v0.1.4) - 2022-10-15

<small>[Compare with v0.1.3](https://github.com/DataShades/ckanext-saml/compare/v0.1.3...v0.1.4)</small>

### Bug Fixes


ckanext/saml/config.py
from __future__ import annotations
from typing import Optional

import ckan.plugins.toolkit as tk

CONFIG_ERROR_TPL = "ckanext.saml.error_template"

CONFIG_SSO_PATH = "ckanext.saml.sso_path"
DEFAULT_SSO_PATH = "/sso/post"

CONFIG_SLO_PATH = "ckanext.saml.slo_path"
DEFAULT_SLO_PATH = "/slo/post"

CONFIG_DYNAMIC = "ckanext.saml.settings.dynamic"
DEFAULT_DYNAMIC = False

CONFIG_USE_REMOTE_IDP = "ckanext.saml.metadata.remote_idp"
DEFAULT_USE_REMOTE_IDP = False

CONFIG_STATIC_HOST = "ckanext.saml.static_host"
DEFAULT_STATIC_HOST = None

CONFIG_USE_FORWARDED_HOST = "ckanext.saml.use_forwarded_host"
DEFAULT_USE_FORWARDED_HOST = False

CONFIG_UNCONDITIONAL_LOGIN = "ckanext.saml.unconditional_login"
DEFAULT_UNCONDITIONAL_LOGIN = False

CONFIG_LOGIN_TEXT = "ckanext.saml.login_button_text"
LEGACY_CONFIG_LOGIN_TEXT = "ckan.saml_login_button_text"
DEFAULT_LOGIN_TEXT = "SAML Login"

CONFIG_REACTIVATE = "ckanext.saml.reactivate_deleted_account"
DEFAULT_REACTIVATE = False

CONFIG_FOLDER_PATH = "ckanext.saml.metadata.base_path"
LEGACY_CONFIG_FOLDER_PATH = "ckan.saml_custom_base_path"
DEFAULT_FOLDER_PATH = "/etc/ckan/default/saml"

CONFIG_HTTPS = "ckan.saml_use_https"
DEFAULT_HTTPS = "off"

CONFIG_USE_NAMEID_AS_EMAIL = "ckan.saml_use_nameid_as_email"
DEFAULT_USE_NAMEID_AS_EMAIL = False

CONFIG_TTL = "ckanext.saml.session.ttl"
DEFAULT_TTL = 30 * 24 * 3600

CONFIG_NAME_FROM_RESPONSE = "ckan.saml.name_from_response"
DEFAULT_NAME_FROM_RESPONSE = False

CONFIG_USER_FIELDS_TRIGGER_UPDATE = "ckan.saml.user_fields_trigger_update"
DEFAULT_USER_FIELDS_TRIGGER_UPDATE = "fullname"


def reactivate_deleted_account() -> bool:
    return tk.asbool(tk.config.get(CONFIG_REACTIVATE, DEFAULT_REACTIVATE))


def sso_path() -> str:
    return tk.config.get(CONFIG_SSO_PATH, DEFAULT_SSO_PATH)


def slo_path() -> str:
    return tk.config.get(CONFIG_SLO_PATH, DEFAULT_SLO_PATH)


def error_template() -> Optional[str]:
    return tk.config.get(CONFIG_ERROR_TPL)


def login_button_text() -> str:
    legacy = tk.config.get(LEGACY_CONFIG_LOGIN_TEXT)
    if legacy:
        return legacy

    return tk.config.get(CONFIG_LOGIN_TEXT, DEFAULT_LOGIN_TEXT)


def folder_path() -> str:
    legacy = tk.config.get(LEGACY_CONFIG_FOLDER_PATH)
    if legacy:
        return legacy

    return tk.config.get(CONFIG_FOLDER_PATH, DEFAULT_FOLDER_PATH)


def use_remote_idp() -> bool:
    return tk.asbool(
        tk.config.get(CONFIG_USE_REMOTE_IDP, DEFAULT_USE_REMOTE_IDP)
    )


def use_dynamic_config() -> bool:
    return tk.asbool(tk.config.get(CONFIG_DYNAMIC, DEFAULT_DYNAMIC))


def unconditional_login() -> bool:
    return tk.asbool(
        tk.config.get(CONFIG_UNCONDITIONAL_LOGIN, DEFAULT_UNCONDITIONAL_LOGIN)
    )


def use_forwarded_host() -> bool:
    return tk.asbool(
        tk.config.get(CONFIG_USE_FORWARDED_HOST, DEFAULT_USE_FORWARDED_HOST)
    )


def static_host() -> Optional[str]:
    return tk.config.get(CONFIG_STATIC_HOST, DEFAULT_STATIC_HOST)


def https() -> str:
    return tk.config.get(CONFIG_HTTPS, DEFAULT_HTTPS)


def use_nameid_as_email() -> bool:
    return tk.asbool(
        tk.config.get(CONFIG_USE_NAMEID_AS_EMAIL, DEFAULT_USE_NAMEID_AS_EMAIL)
    )


def use_name_from_response() -> bool:
    return tk.asbool(
        tk.config.get(CONFIG_NAME_FROM_RESPONSE, DEFAULT_NAME_FROM_RESPONSE)
    )


def user_fields_trigger_update() -> list:
    return tk.config.get(CONFIG_USER_FIELDS_TRIGGER_UPDATE, DEFAULT_USER_FIELDS_TRIGGER_UPDATE).split()
